V1.0.0 RC (#182)
OWASP/maswe: weaknesses/MASVS-RESILIENCE/MASWE-0053.md

Reviewer: DONE

## Modes of Introduction

- **Obfuscation Without Hardening**: Applying standard obfuscation whose transformations are automatically reversible by publicly available tooling.
- **Effectiveness Never Assessed**: Never attempting deobfuscation with current tooling to measure how long the protection actually holds.

## Impact

- **Bypass of Obfuscation Mechanisms**: Attackers can defeat obfuscation mechanisms and as a result, recover a understandable representation of the code that the obfuscation mechanisms were intended to protect, facilitating reverse engineering.
- **Compromise of Sensitive Data**: Attackers can recover proprietary algorithms and embedded secrets despite the obfuscation, resulting in intellectual property exposure.

## Mitigations

- **Use Anti-Decompilation Constructs**: Include constructs that break or degrade the output of common decompilers and disassemblers.
- **Resist Automated Deobfuscation**: Choose control-flow and data transformations that current automated deobfuscators cannot normalize cheaply.

> - **Detect Tampering with Obfuscated Code**: Add self-checks that detect modification or unpacking of obfuscated code and trigger a response (see @MASWE-0064).

Reviewer 2: ```suggestion
```

---
title: Anti-Deobfuscation Techniques Not Implemented
id: MASWE-0053
alias: anti-deobfuscation

> requirement: "The app implements anti-deobfuscation techniques to protect its obfuscation."

Reviewer: Maybe "defend against reverse engineering" instead of "to protect its obfuscation"? Seems more clear to me.

Reviewer 2: Makes sense :)

Reviewer: @TheDauntless I agree with you on applying this change and the other one below :))

platform: [android, ios]
profiles: [R]
threat: MAS-THREAT-0053
attacks: [MAS-ATTACK-0001]
mappings:
  masvs-v1: [MSTG-RESILIENCE-12]
  masvs-v2: [MASVS-RESILIENCE-3]
  cwe: [693]
  maswe-beta: [MASWE-0091]
status: new
---

## Overview

> This weakness occurs when an app relies on obfuscation that automated deobfuscation and static-analysis tooling can undo, because no techniques protect the obfuscation itself.

Reviewer: ```suggestion
This weakness occurs when an app relies on obfuscation techniques that can be reversed (e.g., by automated deobfuscation and static-analysis), because no techniques protect the obfuscation itself.
```

> Obfuscation alone is a moving target: publicly available deobfuscators, decompilers, and scripted analyses can strip common transformations automatically. Anti-deobfuscation techniques, such as anti-decompilation constructs, control-flow shapes that defeat common deobfuscators, and self-checks that detect tampering with obfuscated code, raise the cost of reversing the obfuscation. This weakness complements @MASWE-0051 by protecting the obfuscation itself.

Reviewer: ```suggestion
Publicly available deobfuscators, decompilers, and scripted analyses can strip common transformations automatically. Anti-deobfuscation techniques, such as anti-decompilation constructs and control-flow shapes that defeat common deobfuscators, raise the cost of reversing the obfuscation. This weakness complements @MASWE-0051 by focusing on cases where improperly implemented or insufficiently protected obfuscation can be deobfuscated.
```

Reviewer: This weakness complements @MASWE-0051 by focusing on cases where improperly implemented or insufficiently protected obfuscation can be deobfuscated. I prefer to say that it complements other weakness focusing on other weak cases.

## Modes of Introduction

- **Obfuscation Without Hardening**: Applying standard obfuscation whose transformations are automatically reversible by publicly available tooling.

> - **Deobfuscator-Friendly Control Flow**: Using control-flow patterns that common deobfuscators normalize without difficulty.

Reviewer: ```suggestion
```

Reviewer: This is basically repeating the previous bullet point

> - **No Tamper Detection for Obfuscated Code**: Not detecting when obfuscated code has been modified, unpacked, or re-written by analysis tooling.

Reviewer: ```suggestion
```

Reviewer: This is more of 0064, integrity verification.

- **Effectiveness Never Assessed**: Never attempting deobfuscation with current tooling to measure how long the protection actually holds.

## Impact

> - **Bypass of Protection Mechanisms**: Attackers can recover near-original code and then locate and defeat the app's client-side defenses, resulting in the circumvention of protections the obfuscation was meant to shield.

Reviewer: ```suggestion
- **Bypass of Protection Mechanisms**: Attackers can defeat obfuscation mechanisms and as a result, recover a understandable representation of the code that the obfuscation mechanisms were intended to protect, facilitating reverse engineering.
```

- **Compromise of Sensitive Data**: Attackers can recover proprietary algorithms and embedded secrets despite the obfuscation, resulting in intellectual property exposure.

## Mitigations

- **Use Anti-Decompilation Constructs**: Include constructs that break or degrade the output of common decompilers and disassemblers.
- **Resist Automated Deobfuscation**: Choose control-flow and data transformations that current automated deobfuscators cannot normalize cheaply.

> - **Detect Tampering with Obfuscated Code**: Add self-checks that detect modification or unpacking of obfuscated code and trigger a response (see @MASWE-0064).

Reviewer: As I commented previously, This is more of a mitigation performed on 0064, although this mitigation also protects against this weakness.
